cli/src/commands/rebase.rs

```diff
@@ -443,6 +462,160 @@ fn rebase_revision(
+    for commit in target_commits.iter().rev() {
+        let mut new_parents = IndexSet::new();
+        for old_parent in commit.parent_ids() {
+            if let Some(parents) = new_child_parents.get(old_parent) {
+                new_parents.extend(parents.iter().cloned());
+            } else {
+                new_parents.insert(old_parent.clone());
+            }
+        }
+        new_child_parents.insert(commit.id().clone(), new_parents);
+    }
+
+    let mut rebased_commit_ids = HashMap::new();
+
+    // TODO(ilyagr): Consider making it possible for these descendants
+    // to become emptied, like --skip-empty. This would require writing careful
+    // tests.
+    tx.mut_repo().transform_descendants(
+        settings,
+        target_commits.iter().ids().cloned().collect_vec(),
+        |mut rewriter| {
+            let old_commit = rewriter.old_commit();
+            let old_commit_id = old_commit.id().clone();
+
+            // Commits in the target set should persist only rebased parents from the target
+            // sets.
+            if let Some(new_parents) = new_target_parents.get(&old_commit_id) {
+                // If the commit does not have any parents in the target set, its parents
+                // will be persisted since it is one of the roots of the set.
+                if !new_parents.is_empty() {
+                    rewriter.set_new_rewritten_parents(new_parents.clone());
+                }
+            }
+            // Commits outside the target set should have references to commits inside the set
+            // replaced.
+            else if rewriter
+                .old_commit()
+                .parent_ids()
+                .iter()
+                .any(|id| new_child_parents.contains_key(id))
+            {
+                let mut new_parents = vec![];
+                for parent in rewriter.old_commit().parent_ids() {
+                    if let Some(parents) = new_child_parents.get(parent) {
+                        new_parents.extend(parents.iter().cloned());
+                    } else {
+                        new_parents.push(parent.clone());
+                    }
+                }
+                rewriter.set_new_rewritten_parents(new_parents);
+            }
+            if rewriter.parents_changed() {
+                let builder = rewriter.rebase(settings)?;
+                let commit = builder.write()?;
+                rebased_commit_ids.insert(old_commit_id, commit.id().clone());
+            }
+            Ok(())
+        },
+    )?;
+
+    // Compute the roots of `target_commits`, and update them to account for the
+    // rebase of all of `target_commits`'s descendants.
+    let target_roots: HashSet<_> = new_target_parents
+        .iter()
+        .filter_map(|(commit_id, parents)| {
+            if parents.is_empty() {
+                Some(get_possibly_rewritten_commit_id(
+                    &rebased_commit_ids,
+                    commit_id,
+                ))
```

Comment:
Do we need remapping here? I don't think root commits would be affected by the descendant rebasing.

Reply:
Updated.